concept/thing/RelationImpl.java

```diff
@@ -25,119 +25,70 @@
-import static com.vaticle.typedb.client.concept.type.RoleTypeImpl.protoRoleType;
-import static com.vaticle.typedb.client.concept.type.TypeImpl.protoTypes;
-import static java.util.Arrays.asList;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_add_role_player;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_get_players_by_role_type;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_get_relating;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_get_role_players;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_get_type;
+import static com.vaticle.typedb.client.jni.typedb_client.relation_remove_role_player;
+import static com.vaticle.typedb.client.jni.typedb_client.role_player_get_player;
+import static com.vaticle.typedb.client.jni.typedb_client.role_player_get_role_type;
 
 public class RelationImpl extends ThingImpl implements Relation {
 
-    private final RelationTypeImpl type;
-
-    RelationImpl(String iid, boolean isInferred, RelationTypeImpl type) {
-        super(iid, isInferred);
-        this.type = type;
+    public RelationImpl(com.vaticle.typedb.client.jni.Concept concept) {
+        super(concept);
     }
 
-    public static RelationImpl of(ConceptProto.Thing protoThing) {
-        return new RelationImpl(Bytes.bytesToHexString(protoThing.getIid().toByteArray()),
-                protoThing.getInferred(),
-                RelationTypeImpl.of(protoThing.getType()));
+    @Override
+    public RelationTypeImpl getType() {
+        return new RelationTypeImpl(relation_get_type(nativeObject));
     }
 
     @Override
-    public RelationImpl.Remote asRemote(TypeDBTransaction transaction) {
-        return new RelationImpl.Remote(transaction, getIID(), isInferred(), type);
+    public void addPlayer(TypeDBTransaction transaction, RoleType roleType, Thing player) {
+        relation_add_role_player(((ConceptManagerImpl) transaction.concepts()).transaction,
+                nativeObject, ((RoleTypeImpl) roleType).nativeObject, ((ThingImpl) player).nativeObject);
     }
 
     @Override
-    public RelationTypeImpl getType() {
-        return type;
+    public void removePlayer(TypeDBTransaction transaction, RoleType roleType, Thing player) {
+        relation_remove_role_player(((ConceptManagerImpl) transaction.concepts()).transaction,
+                nativeObject, ((RoleTypeImpl) roleType).nativeObject, ((ThingImpl) player).nativeObject);
     }
 
     @Override
-    public final RelationImpl asRelation() {
-        return this;
+    public Stream<ThingImpl> getPlayers(TypeDBTransaction transaction, RoleType... roleTypes) {
+        return relation_get_players_by_role_type(((ConceptManagerImpl) transaction.concepts()).transaction,
+                nativeObject, Arrays.stream(roleTypes).map(rt -> ((RoleTypeImpl) rt).nativeObject).toArray(com.vaticle.typedb.client.jni.Concept[]::new)).stream().map(ThingImpl::of);
```

Comment:
So the Rust side of this is a "ConceptIterator", what does our SWIG layer give us, a Java iterator?

Reply:
Yep.